src/gt4py/next/ffront/stages.py

```diff
@@ -42,3 +89,84 @@ class PastClosure:
     grid_type: Optional[common.GridType]
     args: tuple[Any, ...]
     kwargs: dict[str, Any]
+
+
+def fingerprint_stage(obj: Any, algorithm: Optional[str | xtyping.HashlibAlgorithm] = None) -> str:
+    hasher: xtyping.HashlibAlgorithm
+    if not algorithm:
+        hasher = xxhash.xxh64()
+    elif isinstance(algorithm, str):
+        hasher = hashlib.new(algorithm)
+    else:
+        hasher = algorithm
+
+    add_content_to_fingerprint(obj, hasher)
+    return hasher.hexdigest()
+
+
+@functools.singledispatch
+def add_content_to_fingerprint(obj: Any, hasher: xtyping.HashlibAlgorithm) -> None:
+    hasher.update(str(obj).encode())
+
+
+@add_content_to_fingerprint.register(FieldOperatorDefinition)
+@add_content_to_fingerprint.register(FoastOperatorDefinition)
+@add_content_to_fingerprint.register(FoastWithTypes)
+@add_content_to_fingerprint.register(FoastClosure)
+@add_content_to_fingerprint.register(ProgramDefinition)
+@add_content_to_fingerprint.register(PastProgramDefinition)
+@add_content_to_fingerprint.register(PastClosure)
+def add_content_to_fingerprint_stages(obj: Any, hasher: xtyping.HashlibAlgorithm) -> None:
+    add_content_to_fingerprint(obj.__class__, hasher)
+    for field in dataclasses.fields(obj):
+        add_content_to_fingerprint(getattr(obj, field.name), hasher)
+
+
+@add_content_to_fingerprint.register
+def add_str_to_fingerprint(obj: str, hasher: xtyping.HashlibAlgorithm) -> None:
+    hasher.update(str(obj).encode())
+
+
+@add_content_to_fingerprint.register(int)
+@add_content_to_fingerprint.register(bool)
+@add_content_to_fingerprint.register(float)
+def add_builtin_to_fingerprint(
+    obj: None,
+    hasher: xtyping.HashlibAlgorithm,
+) -> None:
+    hasher.update(str(obj).encode())
```

Comment:
A suggestion to reduce boilerplate:

```suggestion
@functools.singledispatch
def add_content_to_fingerprint(obj: Any, hasher: xtyping.HashlibAlgorithm) -> None:
    hasher.update(str(obj).encode())

_default_impl = add_content_to_fingerprint[object]
for t in (str, int, bool, float):
    add_content_to_fingerprint.register(t, _default_impl)

@add_content_to_fingerprint.register(FieldOperatorDefinition)
@add_content_to_fingerprint.register(FoastOperatorDefinition)
@add_content_to_fingerprint.register(FoastWithTypes)
@add_content_to_fingerprint.register(FoastClosure)
@add_content_to_fingerprint.register(ProgramDefinition)
@add_content_to_fingerprint.register(PastProgramDefinition)
@add_content_to_fingerprint.register(PastClosure)
def add_content_to_fingerprint_stages(obj: Any, hasher: xtyping.HashlibAlgorithm) -> None:
    add_content_to_fingerprint(obj.__class__, hasher)
    for field in dataclasses.fields(obj):
        add_content_to_fingerprint(getattr(obj, field.name), hasher)
```

Reply:
It was necessary for str and int, but only to avoid a max recursion depth error. I am not exactly sure why this helps.

Reply:
For `str` is needed because otherwise it will call the implementation for `Iterable`, not the default implementation, since `str` is `Iterable` (and that's why the infinite recursion appears). For `int`, I don't see why it would be needed. Are you sure that just `str` is not enough? Maybe I'm misunderstanding something...